### PR TITLE
HHH-14146 Add Presto dialect

### PR DESCRIPTION
**Pull request by findepi**
Add dialect for Presto (https://prestosql.io/)

### CODE REVIEW

**Comment by Sanne** [comment]
Thanks! Great to see support for new modern databases being contributed.

We'll need to add Presto support to our continuous integration systems; my preference would be a container, as it helps us to diagnose issues locally - should we need to.
Would you suggest we use https://github.com/prestosql/presto/tree/master/docker ?

File: hibernate-core/src/main/java/org/hibernate/dialect/PrestoDialect.java [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+		}
+
+		@Override
+		public Connection connect(String url, Properties info) throws SQLException {
+			if ( info != null ) {
+				info = copy( info );
+				// DdlTransactionIsolatorNonJtaImpl works with Presto driver only in autocommit mode (hibernate.connection.autocommit=true)
+				// When this is set, DriverConnectionCreator will pass autocommit connection property which Presto
+				// driver does not accept. Here we remove the property as a workaround, it's being handled on the
+				// Hibernate's side.
```
Comment (Sanne) [inline hibernate-core/src/main/java/org/hibernate/dialect/PrestoDialect.java:111]: I'm a bit concerned about this, could you expand on "it's being handled" ?

What are the implications of this, and what should users do when configuring a production connection pool and/or a Datasource configuration managed by an application server?

**Comment by findepi** [comment]
> We'll need to add Presto support to our continuous integration systems; my preference would be a container, as it helps us to diagnose issues locally - should we need to.
> Would you suggest we use https://github.com/prestosql/presto/tree/master/docker ?

@Sanne correct. You can bring up Presto container  up with eg

```
docker run --rm --name presto -it -p 8080:8080 prestosql/presto:340
```

File: hibernate-core/src/main/java/org/hibernate/dialect/PrestoDialect.java (hunk 1 above)
Comment (findepi) [reply hibernate-core/src/main/java/org/hibernate/dialect/PrestoDialect.java:111]: Currently the `hibernate.connection.autocommit` mode is needed (at least for DDL in tests).
AFAIU this setting controls Hibernate behavior and also is passed to JDBC driver in the form of `autocommit` connection property.
The connection property can be safely ignored (Hibernate uses JDBC APIs to control autocommit of a connection), but Presto JDBC driver does not ignore connection properties that it does not understand.

**Comment by findepi** [comment]
Rebased.

**Comment by findepi** [comment]
The CI build is green.

**Comment by findepi** [comment]
@Sanne can i do anything to help move this forward?

**Comment by findepi** [comment]
@Sanne should i rebase?

**Comment by Sanne** [comment]
hi @findepi ! Apologies, I've been overwhelmed by notifications and missed this completely.

Yes could you rebase it once again please? I'll see to find someone to setup the CI task on CI.

**Comment by findepi** [comment]
Rebased.

There are some build failures, like
```
2020-12-11T16:19:18.4064992Z > Configure project :hibernate-agroal
2020-12-11T16:19:18.4133623Z Unable to decrypt Maven password using PlexusCipher
2020-12-11T16:19:18.4136325Z org.sonatype.plexus.components.cipher.PlexusCipherException: java.lang.ArrayIndexOutOfBoundsException
2020-12-11T16:19:18.4144129Z 	at org.sonatype.plexus.components.cipher.PBECipher.decrypt64(PBECipher.java:193)
2020-12-11T16:19:18.4147647Z 	at org.sonatype.plexus.components.cipher.DefaultPlexusCipher.decrypt(DefaultPlexusCipher.java:72)
2020-12-11T16:19:18.4151391Z 	at org.sonatype.plexus.components.cipher.DefaultPlexusCipher.decryptDecorated(DefaultPlexusCipher.java:86)
2020-12-11T16:19:18.4158323Z 	at org.hibernate.build.publish.auth.maven.pwd.DecryptionPasswordStrategy.decrypt(DecryptionPasswordStrategy.java:78)
```

@Sanne please help me understand what I am doing wrong.

**Comment by Sanne** [comment]
Thanks! Don't worry for the CI job, these are new jobs which have just been added and we still need to sort out some details - sorry about that.

**Comment by Sanne** [comment]
Hi @findepi , with CI having issues I tried to run tests locally - but it failed.

I used:

```
podman run --rm --name presto -it -p 8080:8080 prestosql/presto:340
```

```
./gradlew matrix_presto
```

Did you test it differently, or perhaps some more changes are needed?

**Comment by findepi** [comment]
I didn't run tests recently, only rebased, let me take a look.

**Comment by findepi** [comment]
@Sanne i remember now & sorry for the trouble.

Presto is not capable of housing a typical application, as it still lacks some features commonly seen in relational databases.
For example, if you run `PaginationTest` these are the limitations you would run into:

- Presto allows declaring column to be `NULL` or `NOT NULL`, but the `memory` connector used for testing does not support that yet
- Presto does not support defining indexes (primary keys, unique keys)

For these reasons the preparation part of the test (`PaginationTest#prepareTestData`) won't pass.

As such, you should not expect a typical Hibernate-based app to ever dream of moving to Presto, which is an analytical database.
Yet, some Presto users were asking about Hibernate support for Presto, supposedly to benefit from power of Presto in Java applications without need to write SQL by hand.

See `PrestoDialectTest` for some tests that i added to ensure the dialect actually works.

I also removed the `hibernate-orm/databases/presto/matrix.gradle` file, as i did not find a way to alter test data setup.

**Comment by bradhill99** [comment]
May I know the status of this PR? So there won't be Hibernate support for Presto?
Thanks.

**Comment by beikov** [comment]
As per https://hibernate.atlassian.net/browse/HHH-14693 and https://github.com/hibernate/hibernate-orm/discussions/3783 we decided not to include any dialects in 5.5 anymore for now and focus on 6.0.

IMO it would be fine to add it to 5.5 though if the Dialect for 6 qualifies for hibernate-core according to the requirements document: https://github.com/hibernate/hibernate-orm/blob/wip/6.0/dialects.adoc

**Comment by findepi** [comment]
@bradhill99 if i were to continue here, i would create a Trino dialect, as per https://trino.io/blog/2020/12/27/announcing-trino.html

@beikov i can reopen the PR (with new target module) if needed. 
i closed this, because there was seemingly low interest from potential reviewers, so didn't want to waste your time.

**Comment by beikov** [comment]
You can reopen this PR against the main branch after a Dialect for 6.0 is merged according to the requirements for the hibernate-core inclusion.

**Comment by bradhill99** [comment]
@findepi Thanks for the reply. It will be great to have presto dialect for Hibernate.

**Comment by findepi** [comment]
> after a Dialect for 6.0 is merged according to the requirements for the hibernate-core inclusion.

@beikov is there a PR i should follow?

**Comment by beikov** [comment]
What do you mean by follow? Adding new dialects is a community or vendor driven effort. We as Hibernate team explicitly do not want to bother with any new dialects anymore unless someone steps up and provides full support for that. Read the document about the process: https://github.com/hibernate/hibernate-orm/blob/wip/6.0/dialects.adoc

**Pull request closed without merging**